xptifw/include/xpti_object_table.hpp

```diff
@@ -0,0 +1,155 @@
+#pragma once
+
+#include "spin_lock.hpp"
+
+#include <array>
+#include <functional>
+#include <mutex>
+#include <shared_mutex>
+#include <string_view>
+#include <thread>
+#include <unordered_map>
+#include <variant>
+#include <vector>
+
+namespace xpti {
+// 224 is carefully chosen for Value struct to take 4 cache lines on x86
+template <typename KeyType = uint64_t, size_t SmallSize = 224,
+          size_t SkipSize = 0>
+class ObjectTable {
+public:
+  using HashFunction = std::function<uint64_t(std::string_view)>;
+
+  constexpr static auto DefaultHash = [](std::string_view Data) -> uint64_t {
+    // This is an implementation of FNV hash function.
+    constexpr uint64_t Prime = 1099511628211;
+    uint64_t Hash = 14695981039346656037UL;
+
+    for (char C : Data) {
+      Hash *= Prime;
+      Hash ^= C;
+    }
+
+    return Hash;
```

Comment:
What will happen if a hash collision occurs? Can we address such situation?

Reply:
It should not be a big deal.  We store a bucket of keys for each hash, and we compare values against each candidate in the bucket. We only do it to speed up lookups on insertion. Users are given with an index of the value in a vector, so their lookups are guaranteed to be collision free and O(1).